presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java

```diff
@@ -190,49 +192,54 @@ public AbstractOrcRecordReader(
         // it is possible that old versions of orc use 0 to mean there are no row groups
         checkArgument(rowsInRowGroup > 0, "rowsInRowGroup must be greater than zero");
 
+        long fileRowCount = 0;
+
         // sort stripes by file position
-        List<StripeInfo> stripeInfos = new ArrayList<>();
+        IntArrayList sortedStripeIds = new IntArrayList(fileStripes.size());
         for (int i = 0; i < fileStripes.size(); i++) {
-            Optional<StripeStatistics> stats = Optional.empty();
-            // ignore all stripe stats if too few or too many
-            if (stripeStats.size() == fileStripes.size()) {
-                stats = Optional.of(stripeStats.get(i));
-            }
-            stripeInfos.add(new StripeInfo(fileStripes.get(i), stats));
+            sortedStripeIds.add(i);
+            fileRowCount += fileStripes.get(i).getNumberOfRows();
         }
-        Collections.sort(stripeInfos, comparingLong(info -> info.getStripe().getOffset()));
+        sortedStripeIds.sort((left, right) -> Long.compare(fileStripes.get(left).getOffset(), fileStripes.get(right).getOffset()));
+        this.fileRowCount = fileRowCount;
 
         long totalRowCount = 0;
-        long fileRowCount = 0;
-        ImmutableList.Builder<StripeInformation> stripes = ImmutableList.builder();
+        ImmutableList.Builder<StripeInfo> stripes = ImmutableList.builder();
         ImmutableList.Builder<Long> stripeFilePositions = ImmutableList.builder();
         if (predicate.matches(numberOfRows, getStatisticsByColumnOrdinal(root, fileStats))) {
-            // select stripes that start within the specified split
-            for (StripeInfo info : stripeInfos) {
-                StripeInformation stripe = info.getStripe();
-                if (splitContainsStripe(splitOffset, splitLength, stripe) && isStripeIncluded(root, stripe, info.getStats(), predicate)) {
-                    stripes.add(stripe);
-                    stripeFilePositions.add(fileRowCount);
+            // ignore all stripe stats if too few or too many
+            Optional<List<StripeStatistics>> usableStats = Optional.empty();
+            if (stripeStats.size() == fileStripes.size()) {
+                usableStats = Optional.of(stripeStats);
+            }
+
+            long rowCount = 0;
+            for (int i = 0; i < sortedStripeIds.size(); i++) {
+                // turn loop-index into a final variable to be able to use it in the lambdas
+                final int j = i;
+                StripeInformation stripe = fileStripes.get(sortedStripeIds.getInt(j));
+                Optional<StripeStatistics> stats = usableStats.map(it -> it.get(j));
```

Comment:
`Optional<StripeStatistics> stripeStats = usableStats.map(stats -> stats.get(j));`

Reply:
Stats and stripeStats are already defined in this scope, renamed to stripeStatistics.

`it` is a default name for Kotlin lambdas, Java 8 unfortunately doesn't have this cool feature :(